### PR TITLE
Add node-block and node-return-from for inlining functions with node-return

### PR DESCRIPTION
**Pull request by amorphedstar**
Inlining bodies of functions with return nodes would make it so they returned to the most recent non-inlined function, which is wrong. This PR creates `node-block` and `node-return-from` and implements a traversal to fix this.

### CODE REVIEW

File: src/codegen/transformations.lisp
Comment (amorphedstar) [inline src/codegen/transformations.lisp:143]: Is there an established name for this?

File: src/codegen/transformations.lisp
Comment (amorphedstar) [inline src/codegen/transformations.lisp:149]: Should I make the inliner pass a function name so the block name can be something more informative like `INLINED-FUNCTIONNAME` or `INLINED-LAMBDA`?

File: src/codegen/ast.lisp
Comment (amorphedstar) [inline src/codegen/ast.lisp:270]: Should I rename this slot to `body` ?

Comment (stylewarning) [reply src/codegen/ast.lisp:270]: body makes sense

File: src/codegen/ast.lisp
Comment (stylewarning) [inline src/codegen/ast.lisp:87]: teChnically these read-only things are readers

File: src/codegen/codegen-expression.lisp
Comment (stylewarning) [inline src/codegen/codegen-expression.lisp:249]: nit: newline after the label expression

File: src/codegen/codegen-expression.lisp
Comment (stylewarning) [inline src/codegen/codegen-expression.lisp:252]: nit: newline after the label

File: src/codegen/optimizer.lisp
Comment (stylewarning) [inline src/codegen/optimizer.lisp:403]: add to the comment above, about the returns

File: src/codegen/transformations.lisp
Comment (stylewarning) [inline src/codegen/transformations.lisp:135]: I forgot the semantics of return. Does

    (fn () (return Unit))

work by returning from the nearest enclosed lambda?

File: src/codegen/transformations.lisp
Comment (stylewarning) [reply src/codegen/transformations.lisp:143]: i would say this is about making returns explicit rather than localizing them.

File: src/codegen/transformations.lisp
Comment (stylewarning) [reply src/codegen/transformations.lisp:149]: we might do an earlier pass altogether that transforms ALL functions to always have a block around the contents. thoughts?

**Review by stylewarning — CHANGES_REQUESTED** [review]
generally good. two Q's:

- should we do a single pass that transforms all functions to have a named block with explicated returns?

- does return go back to the nearest enclosing lambda? if so, the logic will need some adjustment.

File: src/codegen/transformations.lisp
Comment (stylewarning) [reply src/codegen/transformations.lisp:135]: Answer: yes, `return` returns from the lambda, so this PR will need adjustments to account for that

File: src/codegen/transformations.lisp
Comment (amorphedstar) [reply src/codegen/transformations.lisp]: I think this is an interesting possibility, and I tried to implement this in the tc-node -> ast-node translation. The sad thing is that the output of the codegen will then look messier with these extra blocks instead, of return-from function-name, unless there is extra logic added to codegen-expression

File: library/system.lisp
Comment (amorphedstar) [inline library/system.lisp:91]: is this right

File: src/typechecker/types.lisp
Comment (amorphedstar) [inline src/typechecker/types.lisp:308]: This would throw type errors like
```lisp
The assertion
(COMMON-LISP:>= COALTON-IMPL/TYPECHECKER/TYPES::NUM
                #1=(COMMON-LISP:LENGTH
                    (COALTON-IMPL/TYPECHECKER/TYPES:FUNCTION-TYPE-ARGUMENTS
                     COALTON-IMPL/TYPECHECKER/TYPES:TY)))
failed with COALTON-IMPL/TYPECHECKER/TYPES::NUM = 1, #1# = 2.
   [Condition of type COMMON-LISP:SIMPLE-ERROR]
```
on code like
```clojure
COALTON-USER> (coalton (let ((f (fn (x) (fn (y) (* x y))))) ((f 1) 2)))
```
or
```clojure
COALTON-USER> (coalton-toplevel (define (f x) (fn (x) (fn (y) (* x y)))))
```

Comment (stylewarning) [reply src/typechecker/types.lisp:308]: @amorphedstar can we add a test for this

File: library/system.lisp
Comment (stylewarning) [reply library/system.lisp:91]: yes

**Review by stylewarning — COMMENTED** [review]
This is looking slick. Is it about ready for merging?

Would be nice to add a simple test for some things, including a test that would trip up the inliner with the returns.

**Comment by amorphedstar** [comment]
> This is looking slick. Is it about ready for merging?
> 
> Would be nice to add a simple test for some things, including a test that would trip up the inliner with the returns.

I added two simple tests, but wasn't quite sure the best place to put them. And I think it should be about ready to merge if you think it's fine having all these extra `block`s now in the codegen output, and potentially unclearer return names in a few places that had special codegen behavior, like `node-bind` in the case when `node-bind-expr` is a `node-abstraction`.

**Comment by stylewarning** [comment]
> > This is looking slick. Is it about ready for merging?
> 
> > 
> 
> > Would be nice to add a simple test for some things, including a test that would trip up the inliner with the returns.
> 
> 
> 
> I added two simple tests, but wasn't quite sure the best place to put them. And I think it should be about ready to merge if you think it's fine having all these extra `block`s now in the codegen output, and potentially unclearer return names in a few places that had special codegen behavior, like `node-bind` in the case when `node-bind-expr` is a `node-abstraction`.

Since blocks are only useful if there's a lexically apparent return, maybe we can (either in this PR or future) only emit a block when needed.

File: src/codegen/codegen-expression.lisp
Comment (stylewarning) [inline src/codegen/codegen-expression.lisp:384]: nit: rename `function-declarations`

File: src/codegen/program.lisp
Comment (stylewarning) [inline src/codegen/program.lisp:130]: nit: indenting looks off

**Review by stylewarning — APPROVED** [review]
I have a few nits that I don't think should hold up the PR. You can address them if desired, but if not within 24h, I'll merge.

We can reduce the number of emitted blocks, but it's fine if you want to
make that a GH issue that can be addressed later. In fact I think
I'd prefer that so as to not let this PR drag.

File: src/codegen/codegen-expression.lisp
Comment (jbouwman) [reply src/codegen/codegen-expression.lisp]: is done

File: src/codegen/program.lisp
Comment (jbouwman) [reply src/codegen/program.lisp]: is done

**Pull request merged**
